src/main/java/com/fasterxml/jackson/databind/deser/std/ReferenceTypeDeserializer.java

```diff
@@ -257,6 +257,6 @@ public Object deserializeWithType(JsonParser p, DeserializationContext ctxt,
         if (_valueTypeDeserializer == null) {
             return deserialize(p, ctxt);
         }
-        return referenceValue(_valueTypeDeserializer.deserializeTypedFromAny(p, ctxt));
+        return typeDeserializer.deserializeTypedFromAny(p, ctxt);
```

Comment:
Wouldn't this change type of deserialized value from `ReferenceType` (like `Optional`) into unwrapped value type? So for `Optional<MyValue>` we would actually deserialize and return `MyValue`, not `Optional.of(MyValue)`.

Which basically would be wrong.

Reply:
That would fail the [integrated test](https://github.com/JooHyukKim/jackson-databind/pull/8/files) (also mentioned above). Lemme check again what typeDeserializer actually does

Reply:
Okay so wrapping with referenceValue() would result having 'Optional.of(Optional.of(value))'

Reply:
Ok something odd is going on then wrt call sequence. I guess `deserializeWithType()` ends up calling plain `deserializeValue()` somehow which does wrapping. Hmmh.

But let's get `AtomicReference` deserialization test added so issue can be more easily demonstrated here.

Reply:
Seems like AtomicReference does not work as well 🤔.
Even wayyy back in 2.8 as [tested via my fork branch PR](https://github.com/JooHyukKim/jackson-databind/pull/9).

Anyhow, let's add test here first.

Reply:
How silly of me... test was wrong 😑.
Trying to deserialize hard-coded JSON `{"ref": 99}` would not work at all.

I will fix AtomicReference test case as well.